### ISSUE TITLE
Uso de entorno, proyectos de prueba.

### ISSUE THREAD

**Issue by josprimen** [josprimen/TFG#4]
Realización de algunos proyectos básicos con el entorno para aprender a usarlo y adquirir facilidad.

**Comment by josprimen** [comment]
Enlaces:
- https://medium.com/@subhamoy.paul986/keras-sequential-api-72e45c39259b (Secuencia de uso Keras)
- https://keras.io/getting-started/sequential-model-guide/ (guía keras modelo secuencial)
- http://www.aprendemachinelearning.com/una-sencilla-red-neuronal-en-python-con-keras-y-tensorflow/ (Red Neuronal sencilla con keras y tensorflow)

- https://github.com/fchollet/deep-learning-with-python-notebooks (Github libro F.Chollet Keras con código, explicaciones y ejemplos)
- https://github.com/keras-team/keras/tree/master/examples (directorio de ejemplos ejericicos keras)

**Comment by josprimen** [comment]
Con el tutorial del primer punto hemos podido saber cual es la secuencia de uso de la librearía keras para crear una red neuronal con esta. 

Con el tutorial del segundo punto hemos creado nuestra primera red neuronal (compuerta XOR), hemos aprendido la forma secuencial, a añadir capas con nodos definidos, compilar, entrenar y ver resultados.
Ademas, hemos aprendido a guardar un modelo entrenado y a cargarlo más tarde.

**Comment by josprimen** [comment]
He tratado de modificar la red neuronal hecha para XOR para predicción de números binarios de cuatro dígitos a decimal.
Nos encontramos con un problema de representación, gran diferencia entre números vecinos como el 7 y 8. 
El uso de Código Gray podría ser una posible solución.

**Comment by josprimen** [comment]
Trabajando en un nuevo ejemplo propuesto en la documentación de Keras relacionado con IMDB.
- https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/3.5-classifying-movie-reviews.ipynb

**Comment by josprimen** [comment]
Preparado el código referente a los datos.

**Comment by josprimen** [comment]
Preparado el código del modelo secuencial y su compilación. Añadidas 2 capas de 16 nodos y una capa de salida con un nodo.

PD: En paralelo al proyecto se está realizando una memoria con explicaciones extendidas, aprendizaje y desarrollo del código.

**Comment by josprimen** [comment]
Realizada la parte de código de entrenamiento y pruebas. 
En el ejemplo junta las dos partes, pero así no funcionaba y decidí separarlas, como en los demás ejemplos realizados.
La ejecución ha sido exitosa.

**Comment by josprimen** [comment]
Buscando información para conectar Pycharm con Github y poder hacer push del código al respositorio actual.

Enlaces:
- https://www.jetbrains.com/help/pycharm/2018.3/manage-projects-hosted-on-github.html?utm_campaign=PY&utm_content=2018.3&utm_medium=link&utm_source=product#clone-from-GitHub
- http://www.drsusansim.org/using-git-and-github-from-pycharm/

Pasos necesarios:

1. Instalación de git en el equipo.
2. En pycharm File>Settings>Version Control>git y añadir el pat del git.exe. En mi caso C:\Program Files\Git\bin\git.exe. OJO: se encuentra en la carpeta bin.
3. En pycharm File>Settings>Version Control>github iniciar sesión con tu cuenta de github
4. En el menú principal clonar el repositorio sobre el que vamos a querer trabajar dándole a VSC>Checkout From Version Control>git y en el diálogo que se abre elegir o añadir la URL del repositorio y aceptar.

Problemas encontrados:
Había seleccionado git-cmd.exe en lugar de git.exe ya que no se encontraba en la ruta que especificaban los tutoriales.

**Comment by josprimen** [comment]
**Tipos de algoritmos**
- Clasificación binaria -> Clasificas en una clase u otra (IMDB crítica positiva o negativa)
- Clasificación multiclase -> Cada ejemplo puede pertenecer a varios grupos ( https://danielmoralesp.gitbooks.io/inteligencia-artificial/content/2-machine-learning/34-coursera-andrew-ng-machine-learning/343-week-3/3433-clasificacion-multiclase.html )
![image](https://user-images.githubusercontent.com/36623141/66391991-e66dc880-e9ce-11e9-810d-52363574050d.png)
- CNN (Convolutional Neural Network) -> Tienen tres dimensiones, no está todo densamente conectado y suele usarse para reconocimiento de imágenes. Tiene una parte que no está densamente conectada y es la encargada de reconocer partes en la imagen y otra que sí lo está y es la encargada de clasificar según esas partes encontradas el objeto. Más info ( https://www.freecodecamp.org/news/an-intuitive-guide-to-convolutional-neural-networks-260c2de0a050/ )
- LSTM (Long short-term memory) -> Las LSTM son un tipo especial de redes recurrentes. La característica principal de las redes recurrentes es que la información puede persistir introduciendo bucles en el diagrama de la red, por lo que, básicamente, pueden «recordar» estados previos y utilizar esta información para decidir cuál será el siguiente. Esta característica las hace muy adecuadas para manejar series cronológicas. Más info ( https://blog.gft.com/es/2018/11/06/como-usar-redes-neuronales-lstm-en-la-prediccion-de-averias-en-las-maquinas/ )
-La regresión consiste en estudiar la relación entre variables y es útil en problemas de predicción.

**Comment by josprimen** [comment]
Normalizar datos ( https://stats.stackexchange.com/questions/70801/how-to-normalize-data-to-0-1-range )



Si tenemos Y que es una matriz de tamaño n filas y m columnas Y.shape devuelve (n,m)
array = geek.arange(8) te crea un array del 0 al 7
array = geek.arange(8).reshape(2, 4) te crea una matriz de 2 filas y 4 columnas con ese array 
[[0 1 2 3]
 [4 5 6 7]]



Para tener los mismos números aleatorios hay que usar una semilla. Ejemplo:

```
import numpy as np
np.random.seed(8)
np.random.rand(3)
Out[4]: array([0.8734294 , 0.96854066, 0.86919454])
np.random.rand(3)
Out[5]: array([0.53085569, 0.23272833, 0.0113988 ])
np.random.seed(8)
np.random.rand(3)
Out[7]: array([0.8734294 , 0.96854066, 0.86919454])

```

**ESTO VALE PARA TENER SIEMPRE EL MISMO VECTOR DE PESOS ALEATORIOS FINAL Y OBTENER LOS MISMO RESULTADOS EN LAS EJECUCIONES.**



Con arrays de numpy si pones conjunto[2:4] te devuelve los arrays que estan en esas posiciones y si pones conjunto[2,0] te devuelve el número que está en la posición 0 del array 2. En este caso solo tenemos un numero en el array, por lo que si ponemos algo que no sea un ,0 nos va a dar error de index.
```
conjunto[2:6]
Out[75]: 
array([[132.],
       [129.],
       [121.],
       [135.]], dtype=float32)
```

**Comment by josprimen** [comment]
**TRATAMIENTO DE DATOS**
drop zeros ->` df.loc[~(df==0).all(axis=1)]`

drop empty rows -> `df.loc[~(df==' ').all(axis=1)]`

**GUARDAR DATOS TRATADOS EN UN CSV**
`dataset.to_csv('pollution.csv')`

**Comment by josprimen** [comment]
**AL HACER ASIGNACIONES NO SE ESTÁ HACIENDO UNA COPIA DEL OBJETO, POR LO QUE SI CAMBIAS LO ASIGNADO, CAMBIA EL VALOR DE LA VARIABLE.**

```
rendimiento = datos[:,0]
acidez = datos[:,1]
datos[:,0] = acidez
```

**AHORA RENDIMIENTO SERÍA IGUAL A LA ACIDEZ**

Solución: `rendimiento = np.copy(datos[:,0])`

**Comment by josprimen** [comment]
**Pasar conjunto de datos en JSON a csv**
With the pandas library, this is as easy as using two commands!

`pandas.read_json()
`To convert a JSON string to a pandas object (either a series or dataframe). Then, assuming the results were stored as df:

`df.to_csv()`
